torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -312,13 +312,24 @@ void RequestCallbackImpl::processRpc(
     case MessageType::SCRIPT_RREF_FETCH_CALL: {
       auto& srf = static_cast<ScriptRRefFetchCall&>(rpc);
       auto& ctx = RRefContext::getInstance();
-      c10::intrusive_ptr<OwnerRRef> rref = ctx.getOwnerRRef(srf.rrefId());
-      if (rref->hasValue()) { // optional fast-path
-        markComplete(ScriptRRefFetchRet({rref->getValue()}).toMessage());
-        return;
-      } else {
+
+      auto futureOwner = ctx.getOwnerRRef(srf.rrefId());
+
+      if (futureOwner->completed()) { // optional fast-path
+        // the OwnerRRef has been created
+        auto& rref = futureOwner->wait();
+        if (rref->hasValue()) {
+          markComplete(ScriptRRefFetchRet({rref->getValue()}).toMessage());
+          return;
+        }
+      }
+
+      futureOwner->addCallback([responseFuture, messageId, futureOwner]() {
```

Comment:
This can now be run in 2 different situations right? When the owner hasn't confirmed the RRef, or when the rref value hasn't been set. Does there need to be any difference in behavior for these 2 situations?

Reply:
> When the owner hasn't confirmed the RRef, or when the rref value hasn't been set. 

This is on owner. So it is actually when 1) the OwnerRRef hasn't been created by the `rpc.remote` message, 2) when the rref value hasn't been set

> This can now be run in 2 different situations right?

This cb is launched when `OwnerRRef` is created. And the nested cb on `whenValueSet` when be launched when the value is ready.

> Does there need to be any difference in behavior for these 2 situations?

Not to my knowledge. They are different cbs. So should be fine?